examples/diagram.dx

```diff
@@ -0,0 +1,193 @@
+
+instance monoidDiagram : Monoid Diagram where
+  mempty = MkDiagram mempty
+  mcombine = \(MkDiagram d1) (MkDiagram d2). MkDiagram $ d1 <> d2
+
+def concatDiagrams (diagrams:n=>Diagram) : Diagram =
+  MkDiagram $ concat for i.
+    (MkDiagram d) = diagrams.i
+    d
+
+-- TODO: arbitrary affine transformations. Our current representation of
+-- rectangles and circles means we can only do scale/flip/rotate90.
+-- Should we use lenses/isomorphisms for these instead?
+def applyTransformation
+      (transformPoint: Point -> Point)
+      (transformGeom:  Geom -> Geom)
+      (d:Diagram) : Diagram =
+  (MkDiagram (AsList _ objs)) = d
+  (MkDiagram $ AsList _ for i.
+    (attr, p, geom) = objs.i
+    (attr, transformPoint p, transformGeom geom))
+
+flipY : Diagram -> Diagram =
+  applyTransformation (\(x,y). (x, -y)) \geom. case geom of
+    Spot          -> Spot
+    Circle r      -> Circle r
+    Rectangle w h -> Rectangle w h
+    Line (x, y)   -> Line (x, -y)
+
+def scale (s:Float) : (Diagram -> Diagram) =
+  applyTransformation ( \(x,y). (s * x, s * y) ) \geom. case geom of
+    Spot          -> Spot
+    Circle r      -> Circle (s * r)
+    Rectangle w h -> Rectangle (s * w) (s * h)
+    Line (x, y)   -> Line (s * x, s * y)
+
+def moveXY ((offX, offY) : Point) : (Diagram -> Diagram) =
+  applyTransformation (\(x,y). (x + offX, y + offY) ) id
+
+def singletonDefault (geom:Geom) : Diagram =
+  MkDiagram $ AsList _ [(defaultAttributes, (0.0, 0.0), geom)]
+
+-- TODO: should really allow nullary functions with `def`
+spot                           : Diagram = singletonDefault Spot
+def circle (r:Float)           : Diagram = singletonDefault $ Circle r
+def rect   (w:Float) (h:Float) : Diagram = singletonDefault $ Rectangle w h
+def line   (p:Point)           : Diagram = singletonDefault $ Line p
+
+def updateAttributes (d:Diagram) (update: Attributes -> Attributes) : Diagram =
+  (MkDiagram (AsList _ objs)) = d
+  MkDiagram $ AsList _ for i.
+    (attr       , geoms) = objs.i
+    (update attr, geoms)
+
+def setFill   (c:Color) (d:Diagram) : Diagram = updateAttributes d $ setAt #fill   (Just c)
+def setStroke (c:Color) (d:Diagram) : Diagram = updateAttributes d $ setAt #stroke (Just c)
+def removeFill          (d:Diagram) : Diagram = updateAttributes d $ setAt #fill   Nothing
+def removeStroke        (d:Diagram) : Diagram = updateAttributes d $ setAt #stroke Nothing
+def setLW     (w:Int)   (d:Diagram) : Diagram = updateAttributes d $ setAt #lw     w
+
+-- === serialization ===
+
+def quote (s:String) : String = str "\"" <> s <> str "\""
+
+def (<+>) (s1:String) (s2:String) : String = s1 <> str " " <> s2
+
+def angleBrackets (s:String) : String = str "<" <> s <> str "/>"
+
+def tagBrackets (tag:String) (s:String) : String =
+  str "<" <> tag <> str ">" <> s <> str "</" <> tag <> str ">"
+
+def tagBracketsAttr (tag:String) (attr:String) (s:String) : String =
+  str "<" <> tag <+> attr <> str ">" <> s <> str "</" <> tag <> str ">"
```

Comment:
or out right string interpolation.

Reply:
Yes, let's add string interpolation! I don't know what the design space looks like but I imagine it's manageable on the implementation side since it's just parsing. Want to open an issue to track it?